ecmascript/dep-graph/src/lib.rs

```diff
@@ -39,46 +37,41 @@ pub struct DependencyDescriptor {
     /// Any leading comments associated with the dependency.  This is used for
     /// further processing of supported pragma that impact the dependency.
     pub leading_comments: Vec<Comment>,
-    /// The location of the import/export statement.
-    pub line: usize,
-    pub col: usize,
+    /// The span of the import/export statement.
+    pub span: Span,
     /// The text specifier associated with the import/export statement.
     pub specifier: JsWord,
-    /// The location of the specifier.
-    pub specifier_line: usize,
-    pub specifier_col: usize,
+    /// The byte position of the start of the specifier.
+    pub specifier_span: Span,
     /// Import assertions for this dependency.
     /// NOTE: it's filled only for static imports and exports.
     pub import_assertions: HashMap<String, String>,
 }
 
-struct DependencyCollector<'a> {
-    comments: &'a SingleThreadedComments,
+struct DependencyCollector<'a, TComments: Comments> {
+    comments: &'a TComments,
     pub items: Vec<DependencyDescriptor>,
-    source_map: &'a SourceMap,
     // This field is used to determine if currently visited "require"
     // is top level and "static", or inside module body and "dynamic".
     is_top_level: bool,
 }
 
-impl<'a> DependencyCollector<'a> {
-    fn get_location(&self, span: Span) -> Loc {
-        self.source_map.lookup_char_pos(span.lo)
-    }
-
+impl<'a, TComments: Comments> DependencyCollector<'a, TComments> {
     fn get_leading_comments(&self, span: Span) -> Vec<Comment> {
-        self.comments
-            .with_leading(span.lo, |comments| comments.to_vec())
+        if let Some(comments) = self.comments.take_leading(span.lo) {
+            let comments_copy = comments.clone();
+            self.comments.add_leading_comments(span.lo, comments);
+            comments_copy
+        } else {
+            Vec::with_capacity(0)
+        }
```

Comment:
This is kind of strange. Perhaps it would be nice to create a `ReadonlyComments` trait that only has the ability to view what's in the comment maps.

Reply:
Actually, thinking about this more we definitely shouldn't do this. This could cause some race conditions in multi-threaded code.

Reply:
If you are not using a trait object, you can define some extension methods which hold the lock for the byte position while executing closure.

Reply:
Hmmm... I'm not sure what you mean. How would this code handle it?

Reply:
I mean an extension method in `Comments` with some object-safety guards like `Self: Sized`.
(I'm not sure if it's a breaking change, though. I'll test it)

It accepts a closure (`FnOnce(&[Comment]) -> Ret`)

And implementor of `Comments` can use it to lock during operation.

Reply:
I'm not sure about avoiding locks. `RefCell` is fast enough for almost all usecases.

I decided to use a global hashmap because I didn't want to pass down comments everywhere, but maybe it was a wrong decision.

Reply:
I'm still not sure how to make that work. I have committed a change that adds some `get` methods that clone... it's not ideal though and is a breaking change... please feel free to change it. I also removed the default implementation of `add_pure_comment` as it had a race condition (also fixed the race condition in the Optimizer code).

I just need to fix some compile errors as the CI is failing... whoops...